extensions/aerosol_optics/mo_aerosol_optics.F90

```diff
@@ -0,0 +1,681 @@
+    real(wp),    dimension(size(aero_type,1), size(aero_type,2), size(this%aero_dust_ext,2)) :: &
+!    real(wp),    dimension(size(aero_type,1), size(aero_type,2), this%get_nband()) :: &
+                 atau, ataussa, ataussag
+                ! ltau, ltaussa, ltaussag, itau, itaussa, itaussag
+                ! Optical properties: tau, tau*ssa, tau*ssa*g
+    integer  :: ncol, nlay, nbnd, nbin, nrh
+    integer  :: icol, ilay, ibnd, ibin
+    ! scalars for total tau, tau*ssa
+    real(wp) :: tau, taussa
+
+    ! ----------------------------------------
+    !
+    ! Error checking
+    !
+    ! ----------------------------------------
+
+    error_msg = ''
+    if(.not.(allocated(this%aero_dust_ext))) then
+      error_msg = 'aerosol optics: no data has been initialized'
+      return
+    end if
+
+    ncol = size(aero_type,1)
+    nlay = size(aero_type,2)
+    nbin = size(this%aero_bin_lo,1)
+    nrh = size(this%aero_rh,1)
+!    nbnd = this%get_nband()
+    nbnd = size(this%aero_dust_ext,2)
+
+    write(61,*) 'Aerosol Optics: checking arrays...'
+    !
+    ! Array sizes
+    !
+    if (check_extents) then
+      if(size(aero_type, 1) /= ncol .or. size(aero_type, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_type has wrong extents"
+      if(size(aero_size, 1) /= ncol .or. size(aero_size, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_size has wrong extents"
+      if(size(aero_mass, 1) /= ncol .or. size(aero_mass, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_mass has wrong extents"
+      if(size(relhum,  1) /= ncol .or. size(relhum,  2) /= nlay) &
+        error_msg = "aerosol optics: relhumn has wrong extents"
+      if(optical_props%get_ncol() /= ncol .or. optical_props%get_nlay() /= nlay) &
+        error_msg = "aerosol optics: optical_props have wrong extents"
+      if(error_msg /= "") return
+    end if
+
+    !
+    ! Spectral consistency
+    !
+    if(check_values) then
+      if(.not. this%bands_are_equal(optical_props)) &
+        error_msg = "aerosol optics: optical properties don't have the same band structure"
+      if(optical_props%get_nband() /= optical_props%get_ngpt() ) &
+        error_msg = "aerosol optics: optical properties must be requested by band not g-points"
+      if(error_msg /= "") return
+    end if
+
+    !$acc data copyin(clwp, ciwp, reliq, reice)                         &
+    !$acc      create(ltau, ltaussa, ltaussag, itau, itaussa, itaussag) &
+    !$acc      create(liqmsk,icemsk)
+    !$omp target data map(to:clwp, ciwp, reliq, reice) &
+    !$omp map(alloc:ltau, ltaussa, ltaussag, itau, itaussa, itaussag) &
+    !$omp map(alloc:liqmsk, icemsk)
+    !
+    ! Aerosol mask; don't need aerosol optics if there's no aerosol
+    !
+    !$acc parallel loop gang vector default(none) collapse(2)
+    !$omp target teams distribute parallel do simd collapse(2)
+    write(61,*) 'Aerosol Optics: defining aerosol mask...'
+    do ilay = 1, nlay
```

Comment:
@mjiacono It will likely be more efficient to check this inside the interpolation rather than pre-computing the mask

Reply:
Aeromsk is pre-computed here to follow what is done in the cloud optics extension.  No change made at this time.

Reply:
@mjiacono One thing we've learned since we wrote the cloud optics is that the codes are all memory-bound, meaning that it's more efficient to put a comparison, especially a simple comparison, close to where it's needed, so we don't have to access arrays of size (ncol, nlay) multiple times.

Reply:
@RobertPincus Understood, but aeromsk is needed where it’s created to limit the bounds checking on the incoming aero_size to elements where aerosols are present.  No change made.

Reply:
@mjiacono I don't quite understand. The mask is created by looking at the value of aerosol type indicating 0. The assignment of aerosol properties has to do the same thing. If you have a CASE DEFAULT that sets optical properties to 0 this achieves the same effect but only requires looking at the aerosol type once.

Reply:
@RobertPincus I see where you’re going now, but it throws me when you pull the code in a different direction from the way you (or Dustin) rewrote the cloud_optics extension.  The array aeromsk has been removed from “compute_all_from_table”. However, the bounds check on aero_size has to remain masked.

Reply:
@mjiacono Thanks for the change and agreed that it only make sense to check sizes where there are aerosols defined.  I appreciate the desire to follow the structure of the cloud optics; keep in mind that we have lots of experience esp. in performance since then